### PR TITLE
Refactoring. Spatial transforms moved to geometric files. Better type…

### PR DESCRIPTION
**Pull request by ternaus**
… hints

### CODE REVIEW

File: albumentations/augmentations/geometric/functional.py [hunk 1]
```diff
@@ -112,7 +128,13 @@ def keypoint_rot90(keypoint, factor, rows, cols, **params):
 
 
 @preserve_channel_dim
-def rotate(img, angle, interpolation=cv2.INTER_LINEAR, border_mode=cv2.BORDER_REFLECT_101, value=None):
+def rotate(
+    img: np.ndarray,
+    angle: int,
```
Comment (Dipet) [inline albumentations/augmentations/geometric/functional.py:133]: float

File: albumentations/augmentations/geometric/functional.py [hunk 2]
```diff
@@ -112,7 +128,13 @@ def keypoint_rot90(keypoint, factor, rows, cols, **params):
...
+    interpolation: int = cv2.INTER_LINEAR,
+    border_mode: int = cv2.BORDER_REFLECT_101,
+    value: Optional[Tuple[int, ...]] = None,
```
Comment (Dipet) [inline albumentations/augmentations/geometric/functional.py:136]: I think `Optional[Union[float, Sequence[float]]]` is better

File: albumentations/augmentations/geometric/functional.py [hunk 3]
```diff
@@ -124,15 +146,15 @@ def rotate(img, angle, interpolation=cv2.INTER_LINEAR, border_mode=cv2.BORDER_RE
     return warp_fn(img)
 
 
-def bbox_rotate(bbox, angle, method, rows, cols):
+def bbox_rotate(bbox: BoxType, angle: int, method: str, rows: int, cols: int) -> BoxType:
```
Comment (Dipet) [inline albumentations/augmentations/geometric/functional.py:149]: angle: float

File: albumentations/augmentations/geometric/functional.py [hunk 4]
```diff
@@ -821,3 +843,439 @@ def bbox_piecewise_affine(
...
+def hflip(img: np.ndarray) -> np.ndarray:
+    return np.ascontiguousarray(img[:, ::-1, ...])
+
+
+def hflip_cv2(img: np.ndarray) -> np.ndarray:
+    return cv2.flip(img, 1)
+
+
+@preserve_shape
+def random_flip(img, code):
```
Comment (Dipet) [inline albumentations/augmentations/geometric/functional.py:861]: Need to add typing

File: albumentations/augmentations/geometric/functional.py [hunk 5]
```diff
@@ -821,3 +843,439 @@ def bbox_piecewise_affine(
...
+    return y, x, angle, scale
+
+
+@preserve_channel_dim
+def pad(
+    img: np.ndarray,
+    min_height: int,
+    min_width: int,
+    border_mode: int = cv2.BORDER_REFLECT_101,
+    value: Optional[Tuple[int, ...]] = None,
```
Comment (Dipet) [inline albumentations/augmentations/geometric/functional.py:1053]: I think Optional[Union[float, Sequence[float]]] is better

File: albumentations/augmentations/geometric/functional.py [hunk 6]
```diff
@@ -821,3 +843,439 @@ def bbox_piecewise_affine(
...
+
+@preserve_channel_dim
+def pad_with_params(
+    img: np.ndarray,
+    h_pad_top: int,
+    h_pad_bottom: int,
+    w_pad_left: int,
+    w_pad_right: int,
+    border_mode: int = cv2.BORDER_REFLECT_101,
+    value: Optional[int] = None,
```
Comment (Dipet) [inline albumentations/augmentations/geometric/functional.py:1091]: I think `Optional[Union[float, Sequence[float]]]` is better

File: albumentations/augmentations/geometric/functional.py [hunk 7]
```diff
@@ -821,3 +843,439 @@ def bbox_piecewise_affine(
...
+@preserve_shape
+def optical_distortion(
+    img,
+    k=0,
+    dx=0,
+    dy=0,
+    interpolation=cv2.INTER_LINEAR,
+    border_mode=cv2.BORDER_REFLECT_101,
+    value=None,
+):
```
Comment (Dipet) [inline albumentations/augmentations/geometric/functional.py:1114]: Need to add typing

File: albumentations/augmentations/geometric/functional.py [hunk 8]
```diff
@@ -821,3 +843,439 @@ def bbox_piecewise_affine(
...
+
+@preserve_shape
+def grid_distortion(
+    img,
+    num_steps=10,
+    xsteps=(),
+    ysteps=(),
+    interpolation=cv2.INTER_LINEAR,
+    border_mode=cv2.BORDER_REFLECT_101,
+    value=None,
```
Comment (Dipet) [inline albumentations/augmentations/geometric/functional.py:1154]: Need to add typing

File: albumentations/augmentations/geometric/functional.py [hunk 9]
```diff
@@ -821,3 +843,439 @@ def bbox_piecewise_affine(
...
+@preserve_shape
+def elastic_transform_approx(
+    img,
+    alpha,
+    sigma,
+    alpha_affine,
+    interpolation=cv2.INTER_LINEAR,
+    border_mode=cv2.BORDER_REFLECT_101,
+    value=None,
+    random_state=None,
```
Comment (Dipet) [inline albumentations/augmentations/geometric/functional.py:1219]: Need to add typing

File: albumentations/augmentations/geometric/transforms.py [hunk 10]
```diff
@@ -910,10 +927,494 @@ def apply_to_bbox(
...
+        self.min_height = min_height
+        self.min_width = min_width
+        self.pad_width_divisor = pad_width_divisor
+        self.pad_height_divisor = pad_height_divisor
+        self.position = PadIfNeeded.PositionType(position)
+        self.border_mode = border_mode
+        self.value = value
+        self.mask_value = mask_value
+
+    def update_params(self, params, **kwargs):
```
Comment (Dipet) [inline albumentations/augmentations/geometric/transforms.py:1001]: typing

File: albumentations/augmentations/geometric/transforms.py [hunk 11]
```diff
@@ -910,10 +927,494 @@ def apply_to_bbox(
...
+                "pad_bottom": h_pad_bottom,
+                "pad_left": w_pad_left,
+                "pad_right": w_pad_right,
+            }
+        )
+        return params
+
+    def apply(
+        self, img: np.ndarray, pad_top: int = 0, pad_bottom: int = 0, pad_left: int = 0, pad_right: int = 0, **params
+    ):
```
Comment (Dipet) [inline albumentations/augmentations/geometric/transforms.py:1050]: return type

File: albumentations/augmentations/geometric/transforms.py [hunk 12]
```diff
@@ -910,10 +927,494 @@ def apply_to_bbox(
...
+            pad_bottom,
+            pad_left,
+            pad_right,
+            border_mode=self.border_mode,
+            value=self.value,
+        )
+
+    def apply_to_mask(
+        self, img: np.ndarray, pad_top: int = 0, pad_bottom: int = 0, pad_left: int = 0, pad_right: int = 0, **params
+    ):
```
Comment (Dipet) [inline albumentations/augmentations/geometric/transforms.py:1063]: return type

File: albumentations/augmentations/geometric/transforms.py [hunk 13]
```diff
@@ -910,10 +927,494 @@ def apply_to_bbox(
...
+        self,
+        bbox: BoxType,
+        pad_top: int = 0,
+        pad_bottom: int = 0,
+        pad_left: int = 0,
+        pad_right: int = 0,
+        rows: int = 0,
+        cols: int = 0,
+        **params
+    ):
```
Comment (Dipet) [inline albumentations/augmentations/geometric/transforms.py:1084]: return type

File: albumentations/augmentations/geometric/transforms.py [hunk 14]
```diff
@@ -910,10 +927,494 @@ def apply_to_bbox(
...
+    # skipcq: PYL-W0613
+    def apply_to_keypoint(
+        self,
+        keypoint: KeypointType,
+        pad_top: int = 0,
+        pad_bottom: int = 0,
+        pad_left: int = 0,
+        pad_right: int = 0,
+        **params
+    ):
```
Comment (Dipet) [inline albumentations/augmentations/geometric/transforms.py:1098]: return type

File: albumentations/augmentations/geometric/transforms.py [hunk 15]
```diff
@@ -910,10 +927,494 @@ def apply_to_bbox(
...
+    def apply(self, img, **params):
+        return F.vflip(img)
+
+    def apply_to_bbox(self, bbox, **params):
+        return F.bbox_vflip(bbox, **params)
+
+    def apply_to_keypoint(self, keypoint, **params):
+        return F.keypoint_vflip(keypoint, **params)
+
+    def get_transform_init_args_names(self):
```
Comment (Dipet) [inline albumentations/augmentations/geometric/transforms.py:1173]: typing

File: albumentations/augmentations/geometric/transforms.py [hunk 16]
```diff
@@ -910,10 +927,494 @@ def apply_to_bbox(
...
+        return F.hflip(img)
+
+    def apply_to_bbox(self, bbox, **params):
+        return F.bbox_hflip(bbox, **params)
+
+    def apply_to_keypoint(self, keypoint, **params):
+        return F.keypoint_hflip(keypoint, **params)
+
+    def get_transform_init_args_names(self):
+        return ()
```
Comment (Dipet) [inline albumentations/augmentations/geometric/transforms.py:1205]: typing

File: albumentations/augmentations/geometric/transforms.py [hunk 17]
```diff
@@ -910,10 +927,494 @@ def apply_to_bbox(
...
+        return {"d": random.randint(-1, 1)}
+
+    def apply_to_bbox(self, bbox, **params):
+        return F.bbox_flip(bbox, **params)
+
+    def apply_to_keypoint(self, keypoint, **params):
+        return F.keypoint_flip(keypoint, **params)
+
+    def get_transform_init_args_names(self):
+        return ()
```
Comment (Dipet) [inline albumentations/augmentations/geometric/transforms.py:1240]: typing

File: albumentations/augmentations/geometric/transforms.py [hunk 18]
```diff
@@ -910,10 +927,494 @@ def apply_to_bbox(
...
+        return F.transpose(img)
+
+    def apply_to_bbox(self, bbox, **params):
+        return F.bbox_transpose(bbox, 0, **params)
+
+    def apply_to_keypoint(self, keypoint, **params):
+        return F.keypoint_transpose(keypoint)
+
+    def get_transform_init_args_names(self):
+        return ()
```
Comment (Dipet) [inline albumentations/augmentations/geometric/transforms.py:1266]: typing

File: albumentations/augmentations/geometric/transforms.py [hunk 19]
```diff
@@ -910,10 +927,494 @@ def apply_to_bbox(
...
+        return bbox_returned
+
+    def get_params(self):
+        return {
+            "k": random.uniform(self.distort_limit[0], self.distort_limit[1]),
+            "dx": round(random.uniform(self.shift_limit[0], self.shift_limit[1])),
+            "dy": round(random.uniform(self.shift_limit[0], self.shift_limit[1])),
+        }
+
+    def get_transform_init_args_names(self):
```
Comment (Dipet) [inline albumentations/augmentations/geometric/transforms.py:1338]: typing

File: albumentations/augmentations/geometric/transforms.py [hunk 20]
```diff
@@ -910,10 +927,494 @@ def apply_to_bbox(
...
+        bbox_returned = bbox_from_mask(mask)
+        bbox_returned = F.normalize_bbox(bbox_returned, rows, cols)
+        return bbox_returned
+
+    def get_params(self):
+        stepsx = [1 + random.uniform(self.distort_limit[0], self.distort_limit[1]) for i in range(self.num_steps + 1)]
+        stepsy = [1 + random.uniform(self.distort_limit[0], self.distort_limit[1]) for i in range(self.num_steps + 1)]
+        return {"stepsx": stepsx, "stepsy": stepsy}
+
+    def get_transform_init_args_names(self):
```
Comment (Dipet) [inline albumentations/augmentations/geometric/transforms.py:1419]: typing

File: albumentations/core/transforms_interface.py [hunk 21]
```diff
@@ -11,11 +11,11 @@
...
 from .utils import format_args
 
-__all__ = ["to_tuple", "BasicTransform", "DualTransform", "ImageOnlyTransform", "NoOp"]
+__all__ = ["to_tuple", "BasicTransform", "DualTransform", "ImageOnlyTransform", "NoOp", "BoxType", "KeypointType"]
 
 NumType = Union[int, float, np.ndarray]
-BoxType = Tuple[float, float, float, float]
-KeypointType = Tuple[float, float, float, float]
+BoxType = Union[Tuple[float, float, float, float], Tuple[float, float, float, float, Any]]
+KeypointType = Union[Tuple[float, float, float, float], Tuple[float, float, float, float, Any]]
```
Comment (Dipet) [inline albumentations/core/transforms_interface.py:18]: Maybe to add additional type, for exapmle `KeypointAndLabels`?

Comment (ternaus) [reply albumentations/core/transforms_interface.py:30]: I was thinking about it. But it is rather complicated to use such Type.

**Review by Dipet — APPROVED** [review]

**Pull request merged**